File: week1/142.linked_list_cycle_ii-环形链表ii/142.linked_list_cycle_ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
// 【93-142. 环形链表 II】
// HashMap
 var detectCycle = function(head) {
     let pos = head;
     let visited = new Set();

     while (pos != null) {
         if (visited.has(pos)) {
             return pos;
         } 
         visited.add(pos);
         pos = pos.next;
     }
     return null;
 };


 // slow- fast
 var detectCycle = function(head) {
     if (head == null || head.next == null) {
         return null; 
     }
     
     let slow = head, fast = head;
     while (true) {
         if (fast == null || fast.next == null) {
             return null;
         }
         slow = slow.next;
         fast = fast.next.next;
         if (slow == fast) {
             break;
         }
     }
     let pos = head;
     while (pos != slow) {
         slow = slow.next;
         pos = pos.next;
     }
     return pos;
};

File: Leet/142.linked_list_cycle_ii-环形链表ii/142.linked_list_cycle_ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// Hashmap
func detectCycle(head *ListNode) *ListNode {
	pos := head
	visited := map[*ListNode]struct{}{}
	for pos != nil {
		if _, ok := visited[pos]; ok {
			return pos
		}
		visited[pos] = struct{}{}
		pos = pos.Next
	}
	return nil
}

// slow- fast
func detectCycle(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}
	slow, fast := head, head

	for true {
		if fast == nil || fast.Next == nil {
			return nil
		}
		fast = fast.Next.Next
		slow = slow.Next
		if fast == slow {
			break
		}
	}
	pos := head
	for slow != pos {
		slow = slow.Next
		pos = pos.Next
	}
	return pos
}